applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/adjoint_nonlinear_strain_energy_response_function.cpp

```diff
@@ -0,0 +1,378 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Mahmoud Sesa, https://github.com/mahmoudsesa
+//
+
+// System includes
+
+// External includes
+#include "utilities/variable_utils.h"
+
+// Project includes
+#include "adjoint_nonlinear_strain_energy_response_function.h"
+
+namespace Kratos 
+{
+    AdjointNonlinearStrainEnergyResponseFunction::AdjointNonlinearStrainEnergyResponseFunction(ModelPart& rModelPart, Parameters ResponseSettings)
+    : AdjointStructuralResponseFunction(rModelPart, ResponseSettings)
+    {
+        // initializing mConditionsRHS 
+        for(auto& cond_i : rModelPart.Conditions())
+        {
+            SizeType number_of_nodes = cond_i.GetGeometry().size();
+            SizeType dimension = cond_i.GetGeometry().WorkingSpaceDimension();
+            SizeType vec_size = number_of_nodes * dimension;
+          
+            mConditionsRHS[cond_i.Id()] = ZeroVector(vec_size);
+
+            // This stores a vector of pointers to the model condition as a member variable because I can not use const condition in 
+            mConditions[cond_i.Id()] = rModelPart.pGetCondition(cond_i.Id());
+
+            // initializes a vector that stores the gradient vectors for the last step
+            mResponseGradient_0[cond_i.Id()] = ZeroVector(vec_size);
+        }
+    }
+
+    AdjointNonlinearStrainEnergyResponseFunction::~AdjointNonlinearStrainEnergyResponseFunction()
+    {
+    }
+
+    double AdjointNonlinearStrainEnergyResponseFunction::CalculateValue(ModelPart& rModelPart)
+    {
+        KRATOS_TRY;
+
+        this->CheckForBodyForces(rModelPart);
+
+        ProcessInfo &r_current_process_info = rModelPart.GetProcessInfo();
+        double response_increment_value = 0.0;
+
+        // Check if there adjoint elements, because response function calculation is done for primal analysis
+        KRATOS_ERROR_IF( r_current_process_info.Has(IS_ADJOINT) && r_current_process_info[IS_ADJOINT] )
+        << "Calculate value for strain energy response is only available when using primal elements" << std::endl;
+
+        // sum all elemental strain energy increment values calculated by trapezoidal rule: E = 0.5 * (f_ext_i - f_ext_i-1) * (u_i - u_i-1)
+        Matrix LHS;
+        Vector RHS;
+        Vector disp;
+        Vector external_force;
+        Vector external_force_previous_step;
+
+        Vector disp_previous_step;
+        Vector disp_increment;
+        Vector average_load;
+
+        // TODO Mahmoud: Calculation using the exact value for the external force at the last time step, not just an approximation
+        for (auto& elem_i : rModelPart.Elements())
```

Comment:
omp with reduction for `response_increment_value`?

Reply:
:+1: